miniapps/autodiff/example.hpp

```diff
@@ -0,0 +1,855 @@
+// Copyright (c) 2010-2021, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef ADEXAMPLE_HPP
+#define ADEXAMPLE_HPP
+
+#include "mfem.hpp"
+#include "admfem.hpp"
+#include <memory>
+#include <iostream>
+#include <fstream>
+
+namespace mfem
+{
+
+///Example: Implementation of the energy and the residual
+/// for p-Laplacian problem. Both, the energy and the residual
+/// are evaluated at the integration points for PDE parameters
+/// vparam and state fields (derivatives with respect to x,y,z
+/// and primal field) stored in vector uu.
+template<typename TDataType, typename TParamVector, typename TStateVector,
+         int residual_size, int state_size, int param_size>
+class MyVFunctor
+{
+public:
+   ///The operator returns the first derivative of the energy with respect
+   /// to all state variables. These are set in vector uu and consist of the
+   /// derivatives with respect to x,y,z and the primal field. The derivative
+   /// is stored in vector rr with length equal to the length of vector uu.
+   void operator()(TParamVector &vparam, TStateVector &uu, TStateVector &rr)
+   {
+      MFEM_ASSERT(residual_size==4,"PLaplacianResidual residual_size should be equal to 4!")
```

Comment:
Why is this a template parameter if it must be 4?

Reply:
The size is necessary for GPU implementation. The statement provided double check for debugging purposes.